File: compiler/bootstrap/lib/aeon/instruction.rb
module Aeon
  class Instruction
    attr_reader :name, :arguments, :line, :column

    NAME_MAPPING = {
      set_integer: 0,
      set_float: 1,
      set_string: 2,
      set_object: 3,
      set_array: 4,
      get_integer_prototype: 6,
      get_float_prototype: 7,
      get_string_prototype: 8,
      get_array_prototype: 9,
      get_thread_prototype: 10,
      get_true_prototype: 11,
      get_false_prototype: 12,
      get_method_prototype: 13,
      get_compiled_code_prototype: 14,
      get_true: 15,
      get_false: 16,
      set_local: 17,
      get_local: 18,
      set_literal_const: 19,
      get_literal_const: 20,
      set_literal_attr: 21,
      get_literal_attr: 22,
      set_compiled_code: 23,
      send_literal: 24,
      return: 25,
      goto_if_false: 26,
      goto_if_true: 27,
      goto: 28,
      def_method: 29,
      def_literal_method: 30,
      run_code: 31,
      get_toplevel: 32,
      is_error: 33,
      error_to_string: 34,
      integer_add: 35,
      integer_div: 36,
      integer_mul: 37,
      integer_sub: 38,
      integer_mod: 39,
      integer_to_float: 40,
      integer_to_string: 41,
      integer_bitwise_and: 42,
      integer_bitwise_or: 43,
      integer_bitwise_xor: 44,
      integer_shift_left: 45,
      integer_shift_right: 46,
      integer_smaller: 47,
      integer_greater: 48,
      integer_equals: 49,
      start_thread: 50,
      float_add: 51,
      float_mul: 52,
      float_div: 53,
      float_sub: 54,
      float_mod: 55,
      float_to_integer: 56,
      float_to_string: 57,
      float_smaller: 58,
      float_greater: 59,
      float_equals: 60,
      array_insert: 61,
      array_at: 62,
      array_remove: 63,
      array_length: 64,
      array_clear: 65,
      string_to_lower: 66,
      string_to_upper: 67,
      string_equals: 68,
      string_to_bytes: 69,
      string_from_bytes: 70,
      string_length: 71,
      string_size: 72,
      stdout_write: 73,
      stderr_write: 74,
      stdin_read: 75,
      stdin_read_line: 76,
      file_open: 77,
      file_write: 78,
      file_read: 79,
      file_read_line: 80,
      file_flush: 81,
      file_size: 82,
      file_seek: 83,
      run_file: 84,
      run_file_dynamic: 85,
      send: 86,
      get_self: 87,
      get_binding_prototype: 88,
      get_binding: 89,
      set_const: 90,
      get_const: 91,
      set_attr: 92,
      get_attr: 93,
      literal_const_exists: 94,
      run_literal_code: 95,
      set_prototype: 96,
      get_prototype: 97,
      local_exists: 98,
      get_caller: 99,
      literal_responds_to: 100,
      responds_to: 101,
      literal_attr_exists: 102
    }

    def initialize(name, arguments, line, column)
      @name = name
      @arguments = arguments
      @line = line
      @column = column

      if !line or !column
        raise ArgumentError, 'A line and column number are required'
      end

      @arguments.each do |arg|
        unless arg.is_a?(Fixnum)
          raise TypeError, "arguments must be Fixnums, not a #{arg.class}"
        end
      end
    end

    def name_integer
      NAME_MAPPING.fetch(name)
    end

    def remap_arguments
      @arguments.map! { |arg| yield arg }
    end

    def inspect
      "Instruction(#{name}, #{arguments.inspect}, line: #{line}, column: #{column})"
    end
  end
end

File: vm/src/instruction.rs
//! Virtual Machine Instructions
//!
//! An Instruction contains information about a single instruction such as the
//! type and arguments.

/// Enum containing all possible instruction types.
#[derive(Debug, Clone)]
#[repr(u16)]
pub enum InstructionType {
    SetInteger               = 0,
    SetFloat                 = 1,
    SetString                = 2,
    SetObject                = 3,
    SetArray                 = 4,
    GetIntegerPrototype      = 6,
    GetFloatPrototype        = 7,
    GetStringPrototype       = 8,
    GetArrayPrototype        = 9,
    GetThreadPrototype       = 10,
    GetTruePrototype         = 11,
    GetFalsePrototype        = 12,
    GetMethodPrototype       = 13,
    GetCompiledCodePrototype = 14,
    GetTrue                  = 15,
    GetFalse                 = 16,
    SetLocal                 = 17,
    GetLocal                 = 18,
    SetLiteralConst          = 19,
    GetLiteralConst          = 20,
    SetLiteralAttr           = 21,
    GetLiteralAttr           = 22,
    SetCompiledCode          = 23,
    SendLiteral              = 24,
    Return                   = 25,
    GotoIfFalse              = 26,
    GotoIfTrue               = 27,
    Goto                     = 28,
    DefMethod                = 29,
    DefLiteralMethod         = 30,
    RunCode                  = 31,
    GetToplevel              = 32,
    IsError                  = 33,
    ErrorToString            = 34,
    IntegerAdd               = 35,
    IntegerDiv               = 36,
    IntegerMul               = 37,
    IntegerSub               = 38,
    IntegerMod               = 39,
    IntegerToFloat           = 40,
    IntegerToString          = 41,
    IntegerBitwiseAnd        = 42,
    IntegerBitwiseOr         = 43,
    IntegerBitwiseXor        = 44,
    IntegerShiftLeft         = 45,
    IntegerShiftRight        = 46,
    IntegerSmaller           = 47,
    IntegerGreater           = 48,
    IntegerEquals            = 49,
    StartThread              = 50,
    FloatAdd                 = 51,
    FloatMul                 = 52,
    FloatDiv                 = 53,
    FloatSub                 = 54,
    FloatMod                 = 55,
    FloatToInteger           = 56,
    FloatToString            = 57,
    FloatSmaller             = 58,
    FloatGreater             = 59,
    FloatEquals              = 60,
    ArrayInsert              = 61,
    ArrayAt                  = 62,
    ArrayRemove              = 63,
    ArrayLength              = 64,
    ArrayClear               = 65,
    StringToLower            = 66,
    StringToUpper            = 67,
    StringEquals             = 68,
    StringToBytes            = 69,
    StringFromBytes          = 70,
    StringLength             = 71,
    StringSize               = 72,
    StdoutWrite              = 73,
    StderrWrite              = 74,
    StdinRead                = 75,
    StdinReadLine            = 76,
    FileOpen                 = 77,
    FileWrite                = 78,
    FileRead                 = 79,
    FileReadLine             = 80,
    FileFlush                = 81,
    FileSize                 = 82,
    FileSeek                 = 83,
    RunLiteralFile           = 84,
    RunFile                  = 85,
    Send                     = 86,
    GetSelf                  = 87,
    GetBindingPrototype      = 88,
    GetBinding               = 89,
    SetConst                 = 90,
    GetConst                 = 91,
    SetAttr                  = 92,
    GetAttr                  = 93,
    LiteralConstExists       = 94,
    RunLiteralCode           = 95,
    SetPrototype             = 96,
    GetPrototype             = 97,
    LocalExists              = 98,
    GetCaller                = 99,
    LiteralRespondsTo        = 100,
    RespondsTo               = 101,
    LiteralAttrExists        = 102,
    SetOuterScope            = 103
}

/// Struct for storing information about a single instruction.
#[derive(Clone, Debug)]
pub struct Instruction {
    /// The type of instruction.
    pub instruction_type: InstructionType,

    /// The arguments of the instruction.
    pub arguments: Vec<u32>,

    /// The line from which the instruction originated.
    pub line: u32,

    /// The column from which the instruction originated.
    pub column: u32
}

impl Instruction {
    /// Returns a new Instruction.
    pub fn new(ins_type: InstructionType, arguments: Vec<u32>, line: u32,
               column: u32) -> Instruction {
        Instruction {
            instruction_type: ins_type,
            arguments: arguments,
            line: line,
            column: column
        }
    }

    pub fn arg(&self, index: usize) -> Result<usize, String> {
        self.arguments
            .get(index)
            .cloned()
            .ok_or(format!("undefined instruction argument {} for {:?}", index, self))
            .map(|num| num as usize)
    }
}
